In [1]:
import pandas as pd
from sklearn.preprocessing import LabelEncoder
from sklearn.model_selection import train_test_split
from sklearn.ensemble import RandomForestClassifier
import numpy as np
from sklearn.tree import DecisionTreeClassifier

In [2]:
data = pd.read_csv('P:/python/Company_Data.csv')
data

Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0     9.50        138      73           11         276    120       Bad   42   
1    11.22        111      48           16         260     83      Good   65   
2    10.06        113      35           10         269     80    Medium   59   
3     7.40        117     100            4         466     97    Medium   55   
4     4.15        141      64            3         340    128       Bad   38   
..     ...        ...     ...          ...         ...    ...       ...  ...   
395  12.57        138     108           17         203    128      Good   33   
396   6.14        139      23            3          37    120    Medium   55   
397   7.41        162      26           12         368    159    Medium   40   
398   5.94        100      79            7         284     95       Bad   50   
399   9.71        134      37            0          27    120      Good   49   

     Education Urban   US  
0           17   Yes  Yes  
1           10   Yes  Yes  
2           12   Yes  Yes  
3           14   Yes  Yes  
4           13   Yes   No  
..         ...   ...  ...  
395         14   Yes  Yes  
396         11    No  Yes  
397         18   Yes  Yes  
398         12   Yes  Yes  
399         16   Yes  Yes  

[400 rows x 11 columns]

In [3]:
data1 = data.copy()

In [4]:
data1['Sales'].describe()

count    400.000000
mean       7.496325
std        2.824115
min        0.000000
25%        5.390000
50%        7.490000
75%        9.320000
max       16.270000
Name: Sales, dtype: float64

In [5]:
data1['Sales'] = pd.cut(data1['Sales'], bins = [0,8,16], labels = ['low','high'])
data1

Sales  CompPrice  Income  Advertising  Population  Price ShelveLoc  Age  \
0    high        138      73           11         276    120       Bad   42   
1    high        111      48           16         260     83      Good   65   
2    high        113      35           10         269     80    Medium   59   
3     low        117     100            4         466     97    Medium   55   
4     low        141      64            3         340    128       Bad   38   
..    ...        ...     ...          ...         ...    ...       ...  ...   
395  high        138     108           17         203    128      Good   33   
396   low        139      23            3          37    120    Medium   55   
397   low        162      26           12         368    159    Medium   40   
398   low        100      79            7         284     95       Bad   50   
399  high        134      37            0          27    120      Good   49   

     Education Urban   US  
0           17   Yes  Yes  
1           10   Yes  Yes  
2           12   Yes  Yes  
3           14   Yes  Yes  
4           13   Yes   No  
..         ...   ...  ...  
395         14   Yes  Yes  
396         11    No  Yes  
397         18   Yes  Yes  
398         12   Yes  Yes  
399         16   Yes  Yes  

[400 rows x 11 columns]

In [6]:
data1['Sales'].value_counts()

low     235
high    163
Name: Sales, dtype: int64

In [7]:
le = LabelEncoder()
data1['Sales'] = le.fit_transform(data1['Sales'])
data1['ShelveLoc'] = le.fit_transform(data1['ShelveLoc'])
data1['Urban'] = le.fit_transform(data1['Urban'])
data1['US'] = le.fit_transform(data1['US'])
data1.head()

Sales  CompPrice  Income  Advertising  Population  Price  ShelveLoc  Age  \
0      0        138      73           11         276    120          0   42   
1      0        111      48           16         260     83          1   65   
2      0        113      35           10         269     80          2   59   
3      1        117     100            4         466     97          2   55   
4      1        141      64            3         340    128          0   38   

   Education  Urban  US  
0         17      1   1  
1         10      1   1  
2         12      1   1  
3         14      1   1  
4         13      1   0

In [8]:
x = data1.iloc[:,1:]
y = data1.iloc[:,0]
print(x)
y.head()

     CompPrice  Income  Advertising  Population  Price  ShelveLoc  Age  \
0          138      73           11         276    120          0   42   
1          111      48           16         260     83          1   65   
2          113      35           10         269     80          2   59   
3          117     100            4         466     97          2   55   
4          141      64            3         340    128          0   38   
..         ...     ...          ...         ...    ...        ...  ...   
395        138     108           17         203    128          1   33   
396        139      23            3          37    120          2   55   
397        162      26           12         368    159          2   40   
398        100      79            7         284     95          0   50   
399        134      37            0          27    120          1   49   

     Education  Urban  US  
0           17      1   1  
1           10      1   1  
2           12      1   1  

0    0
1    0
2    0
3    1
4    1
Name: Sales, dtype: int32

In [9]:
x_train, x_test,y_train,y_test = train_test_split(x,y, test_size=0.25,random_state=40)
print(x_train.shape, x_test.shape,y_train.shape,y_test.shape)

(300, 10) (100, 10) (300,) (100,)


### Random Forest Classification

In [10]:
model = RandomForestClassifier(n_estimators = 100, max_features = 'auto')
model.fit(x_train,y_train)

RandomForestClassifier()

In [20]:
model.feature_importances_

array([0.11642246, 0.12253537, 0.11538947, 0.09092247, 0.21890376,
       0.12060104, 0.12676652, 0.05365383, 0.01854539, 0.01625969])

In [12]:
preds = model.predict(x_test)
preds

array([1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0,
       0, 0, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0,
       1, 1, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0, 1, 1, 1,
       1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 0, 1,
       1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1])

In [13]:
# using train test split
np.mean(preds==y_test)

0.87

In [14]:
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
kfold = KFold(n_splits=10)
results_rf = cross_val_score(model, x_train, y_train, cv=kfold)
print(results_rf.mean())

0.7799999999999999


### Bagging classification

In [15]:
from sklearn.ensemble import BaggingClassifier
kfold = KFold(n_splits=10)
cart = DecisionTreeClassifier()
num_trees = 100
model_b = BaggingClassifier(base_estimator=cart, n_estimators=num_trees, random_state=42)
results_b = cross_val_score(model_b, x_train, y_train, cv=kfold)
print(results_b.mean())

0.7866666666666667


### Adaboosting classification

In [16]:
from sklearn.ensemble import AdaBoostClassifier
kfold = KFold(n_splits=10)
model_a = AdaBoostClassifier(n_estimators=10, random_state=42)
results_a = cross_val_score(model_a, x_train, y_train, cv=kfold)
print(results_a.mean())

0.5433333333333332


In [17]:
model_a.fit(x_train,y_train)

AdaBoostClassifier(n_estimators=10, random_state=42)

In [18]:
preds_a = model_a.predict(x_test)
preds_a

array([0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 1,
       0, 1, 1, 1, 0, 0, 1, 1, 0, 1, 0, 0, 1, 1, 0, 0, 1, 1, 1, 1, 1, 0,
       1, 1, 1, 1, 0, 1, 1, 1, 0, 0, 0, 1, 0, 1, 1, 0, 0, 1, 1, 0, 1, 1,
       1, 0, 1, 1, 1, 1, 0, 1, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1, 1, 1, 0, 0,
       1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 0])

In [19]:
np.mean(preds_a==y_test)

0.59

Firstly i have load a dataset then as in the problem statement for finding the high sales i have divided sales into to named as low and high by the help of pandas cut method. After that with the labelencoder method i have done labelencoding for the necessary columns. After that i have divided the data into input and output by naming as x and y. Then by using trian_test_split ihave divided data into xtrian, ytrian, xtest, ytest by giving test = 25% and random state = 40.
###### Random forest
Now i have build model with random forest classifier by giving no.of trees and number of features(columns) i have set features as auto. And i have fit x_trian and y_trian on model. Then i have done predictions on model by giving xtest then i check model accuarcy is 86% and by using kfold and cross val score i got accuarcy 78%
###### Bagging and boosting
And i have a model using bagging and boosting and i have got accuracy for bagging 78% and for boosting 59%